### DISCUSSION TITLE
Integrate Dare-DB in a real project

### DISCUSSION THREAD

**Discussion by dmarro89** [dmarro89/dare-db#49] [Ideas]
Hi,
I'd like to integrate Dare-DB in a real project.
My Idea is to fork from github a simple project (maybe with a user friendly UI) and replace the used DB with dare-db.
What do you think ? Can be a good idea?
Do you have any hints on what project to fork ?

@vdmitriyev @TheInvincibleRalph

**Comment by vdmitriyev** [comment]
Hi,

very nice idea! Having a reference project, which uses a database is always helpful. I would like to suggest to implement `daredb-web-client`  (alternative name `daredb-webui`). As the name suggests, it should be a web-based client for DareDB. The initial version of a client should be as simple as possible. It would be a good start for a real project using a database and should cover multiple needs at the same time: 

* Demonstrate a real project that uses DareDB
* Act as a `DevTool` to work with a database using web-based UI
* Attract people who want to simply use ```docker compose up``` to spin up a database and see it can do
* Demonstrate advantages and disadvantages of DareDB

I could try try to implement an initial draft. Here are technologies to be used for that:

* [Echo](https://echo.labstack.com/) -> as a backend framework (this could be also a standard HTTP server from `Go`)
* [templ](https://github.com/a-h/templ) -> a a templates-engine for rendering HTML
* [Bootstrap 5](https://getbootstrap.com/docs/5.3/getting-started/introduction/) -> as a frontend toolkit
* (optional) [HTMX](https://htmx.org/) -> to implement interaction between browser and DareDB
* `docker` for deployment

@dmarro89 what do you think about it?

**Reply by dmarro89** [reply]
Hi @vdmitriyev,
I think it's a very good idea.
I'll start to think about its implementation, maybe presenting an idea in the week.
Thanks for your contribution!

**Reply by dmarro89** [reply]
Hi @vdmitriyev,
Do you know any project to take a look at? Something similar to the daredb-webui that you were talking about.

Thanks

**Reply by vdmitriyev** [reply]
@dmarro89 hi!

I don't have a project implement in Go in this particular case. However, a following project could be seen as reference example of WebUI:

* https://github.com/mongo-express/mongo-express

**Comment by vdmitriyev** [comment]
@dmarro89 

After I have seen, that you want to use a ready template to create `daredb-ui`, I wanted to let you know an additional information to consider. Once I also have used a ready template ([flask-adminlte](https://github.com/app-generator/flask-adminlte)). Despite I have learned how to properly design backend for such template, on the frontend side I have deleted everything, expect login form and navigation bars (left and top). Thus, ending up deleted more content from a template than adapting. Besides that, I had to debug JS, to make a template work in a browser, after a number of pages had been deleted. Another point to consider – the template you have selected, have not been updated in the last 3 years. And 3 years is a lot in terms of web UI 😊

Unfortunately, I have no suggestion on what I would use today to create a web UI. But if you want, I could add you to one of the projects I have created, but had no time to publish to open source, so you can check how mentioned by me technologies (echo, templ, HTMX) work together.  

Maybe you could also watch the following short video on HTMX and Go: [Not a Single Line of JavaScript has Been Written??](https://www.youtube.com/watch?v=HDNWjuComhM)

P.S. I have to confess, that I am still looking for a proper JS framework to design flexibly and nice web UI, without any needs to install and run `npm` or similar tooling.
